tests/app/main/test_frameworks.py

```diff
@@ -4095,6 +4106,58 @@ def test_banner_on_service_pages_shows_links_to_company_details_and_declaration(
         else:
             self._assert_incomplete_application_banner_not_visible(submissions.get_data(as_text=True))
 
+    @pytest.mark.parametrize(
+        ('copied', 'link_shown'),
+        (
+            ((False, False, False), True),
+            ((True, False, True), True),
+            ((True, True, True), False),
+        )
+    )
+    def test_drafts_list_has_link_to_add_published_services_if_any_services_not_yet_copied(
+        self, count_unanswered, data_api_client, copied, link_shown
+    ):
+        data_api_client.find_services.return_value = {
+            'services': [
+                {'question1': 'answer1', 'copiedToFollowingFramework': copied[0]},
+                {'question2': 'answer2', 'copiedToFollowingFramework': copied[1]},
+                {'question2': 'answer2', 'copiedToFollowingFramework': copied[2]},
+            ],
+        }
+        data_api_client.get_framework.return_value = self.framework(status='open')
+        self.login()
+
+        res = self.client.get('/suppliers/frameworks/g-cloud-7/submissions/scs')
+        doc = html.fromstring(res.get_data(as_text=True))
+        link = doc.xpath(
+            "//*[@id='content']/p[1]/a[normalize-space(string())='View and add your services from G-Cloud\xa07']"
+        )
+
+        data_api_client.find_services.assert_called_once_with(
```

Comment:
I think we're trying to move away from using things like `assert_called_once_with` on mock objects, because it's easy to typo that method call and then have it get swallowed by the mock.

I think a better test is something like `assert data_api_client.find_services.call_args_list == [mock.call(supplier_id=1234, framework='g-cloud-6', lot='scs', status='published')]`

Reply:
Really? I find doing it that way really ugly! Having to use `mock.call` just gets all cluttered. If you use `autospec=True` with the mock (which I haven't here) it'll catch the error no?

If we're moving away from it as a whole, I'm happy to change here though.

Reply:
👋 

A quick summary (as sent to Sam on Slack)
1) easier to avoid typos that pass silently (especially if the mock itself hasn't been set up correctly). Although `mock` now catches typos for anything starting `assert` or `assret`,  if you absent mindedly miss that `assert` prefix off the method, e.g.  `data_api_client.called_once_with('wrong')`, it will magically pass. Better to have the `assert` up front IMHO.
2) it explicitly states ordering of the calls (and any failed assertions will show if the right args are there but in the wrong order - we can always use a set if we don't care) 
3) I find it easier to visually compare with other test cases where the code may result in more than one function call (or no calls).

Agreed it's a bit verbose, but as mocks are often misunderstood (especially for less experienced folks) I think that's a good thing.

Reply:
Thanks Kat, those are all solid points. I will endeavour to follow them in future!